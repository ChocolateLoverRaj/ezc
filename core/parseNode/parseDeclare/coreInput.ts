import coreTypeParsers from '../coreTypeParsers'
import coreKeyWordsToInputFlags from '../parseInputFlags/coreKeyWordsToInputFlags'
import Input from './Input'

const coreInput: Input = {
  typeParsers: coreTypeParsers,
  keyWordsToInputFlags: coreKeyWordsToInputFlags
}

export default coreInput
