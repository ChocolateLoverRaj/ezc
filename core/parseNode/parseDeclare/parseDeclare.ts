import CoreTokensWithData from '../../CoreTokensWithData'
import CoreTokenType from '../../CoreTokenType'
import CoreKeyWord from '../../CoreKeyWord'
import skip from '../../../util/splitAsyncIterator/skip'
import splitAsyncIterator from '../../../util/splitAsyncIterator/splitAsyncIterator'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import parseIdentifier from '../parseIdentifier'
import tryNodeParsers from '../tryNodeParsers'
import TryParseNode from '../TryParseNode'
import Input from './Input'
import parseInput from '../parseInput/parseInput'
import EnumItemWithData from '../../EnumItemWithData'

const parseDeclare = (
  input: Input
): TryParseNode<CoreNodesWithData[CoreNodeType.DECLARE]> => async stream => {
  const splittedIterator = splitAsyncIterator(stream[Symbol.asyncIterator]())
  let parsedTokens = 0

  {
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD)) return
    const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
    if (!(data.enum === CoreKeyWord && data.id === CoreKeyWord.DECLARE)) return
    skip(splittedIterator, 1)
    parsedTokens++
  }

  const returnType = await tryNodeParsers(input.typeParsers)(splittedIterator.asyncIterable)
  if (returnType === undefined) return
  skip(splittedIterator, returnType.length)
  parsedTokens += returnType.length

  const identifier = await parseIdentifier(splittedIterator.asyncIterable)
  if (identifier === undefined) return
  skip(splittedIterator, identifier.length)
  parsedTokens += identifier.length

  {
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD)) {
      return
    }
    const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
    if (!(data.enum === CoreKeyWord && data.id === CoreKeyWord.OPEN_PARENTHESIS)) return
    skip(splittedIterator, 1)
    parsedTokens++
  }

  // Parse all inputs, there could be 0 - Infinity
  const inputs: EnumItemWithData[] = []
  while (true) {
    // We're done if it's )
    {
      const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
      if (done === true) return
      if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD) {
        const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
        if (data.enum === CoreKeyWord && data.id === CoreKeyWord.CLOSE_PARENTHESIS) {
          skip(splittedIterator, 1)
          parsedTokens++
          break
        }
      }
    }

    const parsedInput = await parseInput(input)(splittedIterator.asyncIterable)
    if (parsedInput === undefined) return
    skip(splittedIterator, parsedInput.length)
    parsedTokens += parsedInput.length

    // Skip past ,
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD) {
      const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
      if (data.enum === CoreKeyWord && data.id === CoreKeyWord.COMMA) {
        skip(splittedIterator, 1)
        parsedTokens++
      }
    }

    inputs.push(parsedInput.node)
  }

  return {
    node: {
      type: {
        enum: CoreNodeType,
        id: CoreNodeType.DECLARE
      },
      data: {
        name: identifier.node.data.name,
        inputs,
        returnType: returnType.node
      }
    },
    length: parsedTokens
  }
}

export default parseDeclare
