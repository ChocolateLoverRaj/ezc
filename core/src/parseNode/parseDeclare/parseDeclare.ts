import CoreTokensWithData from '../../CoreTokensWithData'
import CoreTokenType from '../../CoreTokenType'
import CoreKeyWord from '../../CoreKeyWord'
import skip from 'util/dist/splitAsyncIterator/skip.js'
import splitAsyncIterator from 'util/dist/splitAsyncIterator/splitAsyncIterator.js'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import parseIdentifier from '../parseIdentifier'
import tryNodeParsers from '../tryNodeParsers/tryNodeParsers'
import TryParseNode from '../TryParseNode'
import Input from './Input'
import parseInput from '../parseInput/parseInput'
import EnumItemWithData from '../../EnumItemWithData'
import checkKeyWord from '../checkKeyWord'

const parseDeclare = (
  input: Input
): TryParseNode<CoreNodesWithData[CoreNodeType.DECLARE]> => async stream => {
  const splittedIterator = splitAsyncIterator(stream[Symbol.asyncIterator]())
  let length = 0
  const type = {
    enum: CoreNodeType,
    id: CoreNodeType.DECLARE
  }

  {
    const error = await checkKeyWord(
      splittedIterator.asyncIterable[Symbol.asyncIterator](),
      type,
      length,
      'Expected declare',
      { enum: CoreKeyWord, id: CoreKeyWord.DECLARE }
    )
    if (error !== undefined) return error
    skip(splittedIterator, 1)
    length++
  }

  const parseReturnTypeResult =
    await tryNodeParsers(input.typeParsers)(splittedIterator.asyncIterable)
  if (!parseReturnTypeResult.success) {
    return {
      success: false,
      result: {
        type,
        index: length,
        message: 'Expected return type',
        subAttempts: parseReturnTypeResult.result
      }
    }
  }
  skip(splittedIterator, parseReturnTypeResult.result.length)
  length += parseReturnTypeResult.result.length

  const parseIdentifierResult = await parseIdentifier(splittedIterator.asyncIterable)
  if (!parseIdentifierResult.success) {
    return {
      success: false,
      result: {
        type,
        index: length,
        message: 'Expected identifier',
        subAttempts: [parseIdentifierResult.result]
      }
    }
  }
  skip(splittedIterator, parseIdentifierResult.result.length)
  length += parseIdentifierResult.result.length

  {
    const error = await checkKeyWord(
      splittedIterator.asyncIterable[Symbol.asyncIterator](),
      type,
      length,
      'Expected open parenthesis',
      { enum: CoreKeyWord, id: CoreKeyWord.OPEN_PARENTHESIS }
    )
    if (error !== undefined) return error
    skip(splittedIterator, 1)
    length++
  }

  // Parse all inputs, there could be 0 - Infinity
  const inputs: EnumItemWithData[] = []
  for (let firstInput = true; ; firstInput = false) {
    // We're done if it's )
    {
      const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
      if (done === true) {
        return {
          success: false,
          result: {
            type,
            index: length,
            message: 'Expected close parenthesis',
            subAttempts: undefined
          }
        }
      }
      if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD) {
        const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
        if (data.enum === CoreKeyWord && data.id === CoreKeyWord.CLOSE_PARENTHESIS) {
          skip(splittedIterator, 1)
          length++
          break
        }
      }
    }

    // Skip past ,
    if (!firstInput) {
      const error = await checkKeyWord(
        splittedIterator.asyncIterable[Symbol.asyncIterator](),
        type,
        length,
        'Expected comma',
        { enum: CoreKeyWord, id: CoreKeyWord.COMMA }
      )
      if (error !== undefined) return error
      skip(splittedIterator, 1)
      length++
    }

    const parseInputResult = await parseInput(input)(splittedIterator.asyncIterable)
    if (!parseInputResult.success) {
      return {
        success: false,
        result: {
          type,
          index: length,
          message: 'Expected input',
          subAttempts: [parseInputResult.result]
        }
      }
    }
    skip(splittedIterator, parseInputResult.result.length)
    length += parseInputResult.result.length

    inputs.push(parseInputResult.result.node)
  }

  return {
    success: true,
    result: {
      node: {
        type,
        data: {
          name: parseIdentifierResult.result.node,
          inputs,
          returnType: parseReturnTypeResult.result.node
        }
      },
      length
    }
  }
}

export default parseDeclare
