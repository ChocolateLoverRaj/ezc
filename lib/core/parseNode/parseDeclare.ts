import CoreTokensWithData from '../CoreTokensWithData'
import CoreTokenType from '../CoreTokenType'
import KeyWord from '../KeyWord'
import OpenCloseType from '../OpenCloseType'
import skip from '../splitAsyncIterator/skip'
import splitAsyncIterator from '../splitAsyncIterator/splitAsyncIterator'
import CoreNodesWithData from './CoreNodesWithData'
import CoreNodeType from './CoreNodeType'
import InputFlag from './InputFlag'
import InputType from './InputType'
import parseIdentifier from './parseIdentifier'
import tryNodeParsers from './tryNodeParsers'
import TryParseNode from './TryParseNode'
import TypeParser from './TypeParser'

const parseDeclare = (
  typeParsers: readonly TypeParser[]
): TryParseNode<CoreNodesWithData[CoreNodeType.DECLARE]> => async stream => {
  const splittedIterator = splitAsyncIterator(stream[Symbol.asyncIterator]())
  let parsedTokens = 0

  {
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD)) return
    const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
    if (!(data.enum === KeyWord && data.id === KeyWord.DECLARE)) return
    skip(splittedIterator, 1)
    parsedTokens++
  }

  const returnType = await tryNodeParsers(typeParsers)(splittedIterator.asyncIterable)
  if (returnType === undefined) return
  skip(splittedIterator, returnType.length)
  parsedTokens += returnType.length

  const identifier = await parseIdentifier(splittedIterator.asyncIterable)
  if (identifier === undefined) return
  skip(splittedIterator, identifier.length)
  parsedTokens += identifier.length

  {
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.OPEN_CLOSE)) {
      return
    }
    const { data: { type, close } } = value as CoreTokensWithData[CoreTokenType.OPEN_CLOSE]
    if (!(type === OpenCloseType.PARENTHESIS && !close)) return
    skip(splittedIterator, 1)
    parsedTokens++
  }

  // Parse all inputs, there could be 0 - Infinity
  const inputTypes: InputType[] = []
  const keyWordsToInputFlags: Partial<Record<KeyWord, InputFlag>> = {
    [KeyWord.NO_CAPTURE]: InputFlag.NO_CAPTURE,
    [KeyWord.NO_ALIAS]: InputFlag.NO_ALIAS
  }
  while (true) {
    // We're done if it's )
    {
      const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
      if (done === true) return
      if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.OPEN_CLOSE) {
        const { data: { type, close } } = value as CoreTokensWithData[CoreTokenType.OPEN_CLOSE]
        if (type === OpenCloseType.PARENTHESIS && close) {
          skip(splittedIterator, 1)
          parsedTokens++
          break
        }
      }
    }

    const inputType = await tryNodeParsers(typeParsers)(splittedIterator.asyncIterable)
    if (inputType === undefined) return
    skip(splittedIterator, inputType.length)
    parsedTokens += inputType.length

    // Parse all flags
    const flags: InputFlag[] = []
    while (true) {
      const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
      if (done === true) return
      if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD)) break
      const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]

      // We're done if it's ,
      if (data.enum === KeyWord && data.id === KeyWord.COMMA) {
        skip(splittedIterator, 1)
        parsedTokens++
        break
      }

      if (!(data.enum === KeyWord && Object.hasOwn(keyWordsToInputFlags, data.id))) break
      skip(splittedIterator, 1)
      parsedTokens++
      flags.push(keyWordsToInputFlags[data.id])
    }

    inputTypes.push({ type: inputType.node, flags })
  }

  return {
    node: {
      type: {
        enum: CoreNodeType,
        id: CoreNodeType.DECLARE
      },
      data: {
        name: identifier.node.data.name,
        type: {
          returnType: returnType.node,
          inputTypes
        }
      }
    },
    length: parsedTokens
  }
}

export default parseDeclare
