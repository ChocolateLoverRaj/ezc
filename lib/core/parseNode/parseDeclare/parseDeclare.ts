import CoreTokensWithData from '../../CoreTokensWithData'
import CoreTokenType from '../../CoreTokenType'
import CoreKeyWord from '../../CoreKeyWord'
import OpenCloseType from '../../OpenCloseType'
import skip from '../../splitAsyncIterator/skip'
import splitAsyncIterator from '../../splitAsyncIterator/splitAsyncIterator'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import InputType from '../InputType'
import parseIdentifier from '../parseIdentifier'
import tryNodeParsers from '../tryNodeParsers'
import TryParseNode from '../TryParseNode'
import parseInputFlags from '../parseInputFlags/parseInputFlags'
import Input from './Input'

const parseDeclare = (
  { typeParsers, keyWordsToInputFlags }: Input
): TryParseNode<CoreNodesWithData[CoreNodeType.DECLARE]> => async stream => {
  const splittedIterator = splitAsyncIterator(stream[Symbol.asyncIterator]())
  let parsedTokens = 0

  {
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD)) return
    const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
    if (!(data.enum === CoreKeyWord && data.id === CoreKeyWord.DECLARE)) return
    skip(splittedIterator, 1)
    parsedTokens++
  }

  const returnType = await tryNodeParsers(typeParsers)(splittedIterator.asyncIterable)
  if (returnType === undefined) return
  skip(splittedIterator, returnType.length)
  parsedTokens += returnType.length

  const identifier = await parseIdentifier(splittedIterator.asyncIterable)
  if (identifier === undefined) return
  skip(splittedIterator, identifier.length)
  parsedTokens += identifier.length

  {
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.OPEN_CLOSE)) {
      return
    }
    const { data: { type, close } } = value as CoreTokensWithData[CoreTokenType.OPEN_CLOSE]
    if (!(type === OpenCloseType.PARENTHESIS && !close)) return
    skip(splittedIterator, 1)
    parsedTokens++
  }

  // Parse all inputs, there could be 0 - Infinity
  const inputTypes: InputType[] = []
  while (true) {
    // We're done if it's )
    {
      const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
      if (done === true) return
      if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.OPEN_CLOSE) {
        const { data: { type, close } } = value as CoreTokensWithData[CoreTokenType.OPEN_CLOSE]
        if (type === OpenCloseType.PARENTHESIS && close) {
          skip(splittedIterator, 1)
          parsedTokens++
          break
        }
      }
    }

    const inputType = await tryNodeParsers(typeParsers)(splittedIterator.asyncIterable)
    if (inputType === undefined) return
    skip(splittedIterator, inputType.length)
    parsedTokens += inputType.length

    // Parse all flags
    const parsedFlags = await parseInputFlags(keyWordsToInputFlags)(splittedIterator.asyncIterable)
    skip(splittedIterator, parsedFlags.length)
    parsedTokens += parsedFlags.length

    // Skip past ,
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD) {
      const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
      if (data.enum === CoreKeyWord && data.id === CoreKeyWord.COMMA) {
        skip(splittedIterator, 1)
        parsedTokens++
      }
    }

    inputTypes.push({
      type: inputType.node,
      flags: parsedFlags.flags
    })
  }

  return {
    node: {
      type: {
        enum: CoreNodeType,
        id: CoreNodeType.DECLARE
      },
      data: {
        name: identifier.node.data.name,
        type: {
          returnType: returnType.node,
          inputTypes
        }
      }
    },
    length: parsedTokens
  }
}

export default parseDeclare
