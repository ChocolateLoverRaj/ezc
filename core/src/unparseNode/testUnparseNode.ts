import TryParseNode from '../parseNode/TryParseNode'
import EnumItemWithData from '../EnumItemWithData'
import UnparsedPartType from './UnparsedPartType'
import unparsedNodeToString from '../unparsedNodeToString/unparsedNodeToString'
import coreInput from '../unparsedNodeToString/coreInput'
import parseAllTokens from '../tryGetToken/parseAllTokens'
import coreTryers from '../tryGetToken/coreTryers'
import arrayToAsyncIterable from 'util/dist/arrayToAsyncIterable/arrayToAsyncIterable.js'

const testUnparseNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  expect(unparsedNodeToString(coreInput)([{
    type: UnparsedPartType.NODE,
    data: ((await parseNode(
      parseAllTokens(coreTryers)(
        arrayToAsyncIterable([code])) as any))?.result as any).node as EnumItemWithData
  }])).toMatchSnapshot()
}

export default testUnparseNode
