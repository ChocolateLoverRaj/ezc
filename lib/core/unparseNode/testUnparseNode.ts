import TryParseNode from '../parseNode/TryParseNode'
import EnumItemWithData from '../EnumItemWithData'
import UnparsedPartType from './UnparsedPartType'
import unparsedNodeToString from '../unparsedNodeToString/unparsedNodeToString'
import coreInput from '../unparsedNodeToString/coreInput'
import parseAllTokens from '../tryGetToken/parseAllTokens'
import coreTryers from '../tryGetToken/coreTryers'
import arrayToAsyncIterable from '../arrayToAsyncIterable'

const testUnparseNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  expect(unparsedNodeToString(coreInput)([{
    type: UnparsedPartType.NODE,
    data: (await parseNode(
      parseAllTokens(coreTryers)(
        arrayToAsyncIterable([code])) as any))?.node as EnumItemWithData
  }])).toMatchSnapshot()
}

export default testUnparseNode
