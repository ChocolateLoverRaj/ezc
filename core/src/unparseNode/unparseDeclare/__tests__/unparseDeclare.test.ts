import parseDeclare from '../../../parseNode/parseDeclare/parseDeclare'
import coreParseDeclareInput from '../../../parseNode/parseDeclare/coreInput'
import testUnparseNode from '../../testUnparseNode'

test('declare void @someFn(ptr nofree, i1)', async () => {
  await testUnparseNode(parseDeclare(coreParseDeclareInput), 'declare void @someFn(ptr nofree, i1)')
})
