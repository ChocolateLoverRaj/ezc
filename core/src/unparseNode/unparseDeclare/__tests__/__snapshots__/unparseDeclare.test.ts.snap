// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`declare void @someFn(ptr nofree, i1) 1`] = `"declare void @someFn(ptr nofree, i1)"`;
